### ISSUE TITLE
[1.15.2] Invalid packet received when connecting to vanilla server with resource pack

### ISSUE THREAD

**Issue by Project579** [MinecraftForge/MinecraftForge#6688]
**Minecraft Version:** 1.15.2

**Forge Version:** 31.1.70

**Logs:** https://gist.github.com/Project579/a1a9f3d00c3f6eba56b93d2c64a0fd0a

**Steps to Reproduce:**
 1. Connect to a Vanilla Server with custom resource packs;
 2. Wait for connection;
 3. Client will show error on screen: "Invalid packet received, aborting connection";
 4. Client will disconnect from the server.

**Description of issue:**

Since versions after 31.1.0 (31.1.0 is the version I tested, later versions probably worked until 3.1.15) of Minecraft Forge the Client is unable to connect to vanilla servers using custom resource packs, exception can be found at line 1167 of the log (java.lang.IllegalStateException: Invalid packet received, aborting connection) , I also tried running Forge with no mods and the issue remains; 
It seems one of the custom request packets for the resource pack is using an incorrect network direction (vanilla server side bug?).

In short the "NetworkHooks.validatePacketDirection" control added to resolve security issues on servers in commit: https://github.com/MinecraftForge/MinecraftForge/commit/aca45340bfd1ae8de9857cb906a2a31fbf8fbcdb seems to have broken custom resource pack loading.

**Comment by jamierocks** [comment]
I've just tried this myself with a plain Vanilla 1.15.2 server - could not reproduce your issue. Looking at the stacktrace, you have Optifine installed - which could be the cause of this.

![https://i.jamierocks.uk/2TEFDqBI5c.png](https://i.jamierocks.uk/2TEFDqBI5c.png)

> I also tried running Forge with no mods and the issue remains

We need a stacktrace from that environment then.

**Comment by Project579** [comment]
Here is the log as you requested no mods only forge [31.1.87] just installed:

https://gist.github.com/Project579/dae280b65784c2cee1cbd18ef84753cb

The client is able to connect but as soon as the resource pack loading starts the exception appears and the client gets disconnected.

**Comment by DaemonUmbra** [comment]
That’s not a Vanilla server, that’s waterfall. And it appears they send non-vanilla packets to clients, and it just so happens one of them has the same discriminator as a Forge packet that isn’t meant to arrive on that channel.

**Comment by Project579** [comment]
Well I guess then it's about deciding on which side it's better to fix these kinds of situations then, I would assume on Waterfall side?

**Comment by LexManos** [comment]
Cant not reproduce on anything we have control of, So blame Waterfall. If they send invalid packets, they should fix that.
